File: cpp_src/rosbag_test/src/bag_play.cpp
#include <iostream>
#include <ros/ros.h>
#include <rosbag/bag.h>
#include <rosbag/view.h>
#include <topic_tools/shape_shifter.h>
#include "cxxopts.hpp"

bool findTopic(const std::string topic, const std::vector<std::string>& topic_list) {
    if (topic_list.empty()) {
        return true;
    }

    for (auto & item : topic_list) {
        // ROS_INFO("%s %s", topic.c_str(), item.c_str());
        if (topic == item) {
            return true;
        }
    }

    return false;
}

int main(int argc, char ** argv) {
  ros::init(argc, argv, "bag_play");
  ros::NodeHandle nh;

  // https://github.com/jarro2783/cxxopts
  cxxopts::Options options(argv[0], "parse cmd line");
  options.add_options()
      ("b,bag", "name of the bag", cxxopts::value<std::string>())
      ("s,start_offset", "sec offset of start time", cxxopts::value<int>())
      ("r,rate", "player rate", cxxopts::value<double>())
      ("t,topic", "topic list", cxxopts::value<std::vector<std::string>>())
      ("h,help", "show help");
  auto result = options.parse(argc, argv);  

  if (result.count("help")) {
    ROS_INFO("%s", options.help().c_str());
    return 0;
  }

  if (!result.count("bag")) {
    ROS_ERROR("please specify bag using -b");
    return -1;
  }   

  std::string bag_name;
  int start_offset_sec = 0;
  double play_delay_sec = 2.0;
  double play_rate = 1.0;
  std::vector<std::string> topic_list;
  if (result.count("bag")) {
    bag_name = result["bag"].as<std::string>();
  }  
  if (result.count("start_offset")) {
    start_offset_sec = result["start_offset"].as<int>();
  }
  if (result.count("rate")) {
    play_rate = result["rate"].as<double>();
  }  
  if (result.count("topic")) {
      topic_list = result["topic"].as<std::vector<std::string>>();
  } 

  rosbag::Bag bag;
  try {
    bag.open(bag_name.c_str(), rosbag::bagmode::Read);
  } catch (rosbag::BagException& e) {
    ROS_ERROR("open bag failed %s", e.what());
    return -1;
  }

  rosbag::View view(bag);
  std::map<std::string, ros::Publisher> pubs;

  const std::vector<const rosbag::ConnectionInfo*> connections = view.getConnections();
  for (auto& item : connections) {
    if (findTopic(item->topic, topic_list)) {
      std::string topic_name = item->topic;
      ROS_INFO("%s %s %s", topic_name.c_str(), item->datatype.c_str(), item->md5sum.c_str()/*, item->msg_def.c_str()*/);
      ros::AdvertiseOptions opts;
      opts.topic = topic_name;
      opts.queue_size = 1000;
      opts.md5sum = item->md5sum;
      opts.datatype = item->datatype;
      opts.message_definition = item->msg_def;
      
      pubs[topic_name] = nh.advertise(opts);
    }
  }

  ROS_INFO("start play bag after %f sec...", play_delay_sec);
  ros::Duration(0, (int)(play_delay_sec * 1e9)).sleep();

  ros::Time begin_time = view.getBeginTime() + ros::Duration(start_offset_sec, 0);
  ros::Duration stone_delta = ros::Time::now() - begin_time;

  int cnt = 0;
  ros::Duration sum_du = ros::Duration();
  for (const auto& msg : view) {
    if (msg.getTime() >= begin_time) {

      ros::Duration msg_delta = ros::Duration(ros::Duration(msg.getTime() - begin_time).toSec() / play_rate);
      while (begin_time + msg_delta + stone_delta > ros::Time::now()) {
        ros::Duration du = ros::Duration(ros::Duration(msg.getTime() - begin_time).toSec()/play_rate) - sum_du;
        sum_du += du;
        du.sleep();
      }

      printf("[RUNNING] Bag Time: %.6f  Duration: %.6f\r", msg.getTime().toSec(), ros::Duration(msg.getTime() - begin_time).toSec());
      fflush(stdout);

      if (findTopic(msg.getTopic(), topic_list)) {
        topic_tools::ShapeShifter::ConstPtr smsg = msg.instantiate<topic_tools::ShapeShifter>();
        pubs[msg.getTopic()].publish(smsg);
      }

    }

    if (!ros::ok()) {
      break;
    }
  }

  printf("\n");

  bag.close();


  return 0;
}

File: pyhon_src/rosbag_test/scripts/bag_play.py
#! /usr/bin/env python3

import os, sys
import rospy
import rosbag
import argparse

class BagPlay(object):
  def __init__(self, bag, start_offset, rate, topic_list=[]):
    self.__bag = bag
    self.__start_ofset = start_offset
    self.__rate = rate
    self.__topic_list = topic_list
    self.__publisher = {}

  def create_publisher(self):
    bag = rosbag.Bag(self.__bag, 'r')

    for topic, msg, t in bag.read_messages(topics=self.__topic_list, raw=False):
      if topic not in self.__publisher:
        self.__publisher[topic] = rospy.Publisher(topic, type(msg), queue_size=10)

    bag.close()

  def start_play(self):
    bag = rosbag.Bag(self.__bag, "r")

    bag_start_time = bag.get_start_time() + self.__start_ofset
    rospy_start_time =rospy.get_time()
    for topic, msg, t in bag.read_messages(topics=self.__topic_list, start_time=rospy.Time.from_sec(bag_start_time)):
      relative_time = (t.to_sec() - bag_start_time) / self.__rate
      while rospy.get_time() < (rospy_start_time + relative_time):
        rospy.sleep(0.01)

      self.__publisher[topic].publish(msg)

      # print("%f" %t.to_sec())
      cur = t.to_sec()
      du = rospy.Duration(cur - bag_start_time).to_sec()
      sys.stdout.write("[RUNNING] Bag Time: %f Duration: %f\r" %(cur, du))
      sys.stdout.flush()

      if rospy.is_shutdown():
        break

    print("\n")
    bag.close

def main():
  parser = argparse.ArgumentParser(description="play topic from rosbag")
  parser.add_argument("-b", "--bag", type=str, required=True, help="specify rosbag")
  parser.add_argument("-s", "--start_offset", type=int, default=0, help="offset time from bag start")
  parser.add_argument("-r", "--rate", type=float, default=1.0, help="specify play rate")
  parser.add_argument("-t", "--topic_list", nargs="+", type=str, help="specify play topics")
  parser.add_argument("extra_args", nargs='*', help=argparse.SUPPRESS)

  args = parser.parse_args()

  if not os.path.isfile(args.bag):
    print("%s is not found!" %args.bag)
    return

  rospy.init_node("bag_player")

  player = BagPlay(args.bag, args.start_offset, args.rate, args.topic_list)

  player.create_publisher()

  print("start play bag after 2 sec...")
  rospy.sleep(2.0)

  player.start_play()

if __name__ == "__main__":
  main()